File: OJ/leetcode/python/93_restore_ip_addresses.py
from typing import List

class Solution:
    def remainValid(self, group: int, s: str) -> bool:
        if group == 1 and 3 <= len(s) <= 9 or \
            group == 2 and 2 <= len(s) <= 6 or \
            group == 3 and 1 <= len(s) <= 3 or \
            group == 4 and len(s) == 0:
            return True
        else:
            return False

    def restoreIpAddresses(self, s: str) -> List[str]:
        ans = []
        if len(s) < 4 or len(s) > 12:
            return ans
        self.dfs(1, s, "", ans)
        return ans

    def dfs(self, group: int, s: str, ip: str, ans: List[str]) -> None:
        if group > 4:
            ans.append(ip[:(len(ip) - 1)])
            return
        for i in range(1, 4):
            prefix = s[:i]
            if i > len(s):
                break
            if not self.remainValid(group, s[i:]):
                continue
            value = int(prefix)
            if value <= 255 and (value == 0 and i == 1 or prefix[0] != "0"):
                self.dfs(group + 1, s[i:], ip + prefix + ".", ans)


solution = Solution()
print(solution.restoreIpAddresses("1111"))
print(solution.restoreIpAddresses("25525511135"))
print(solution.restoreIpAddresses("0000"))
print(solution.restoreIpAddresses("010010"))
print(solution.restoreIpAddresses("101023"))

File: competitive_programming/leetcode/cpp/93_restore_ip_addresses.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    vector<string> restoreIpAddresses(string s) {
        vector<string> ans;
        if (s.size() > 12 || s.size() < 4) {
            return ans;
        }
        dfs(1, "", s, ans);
        return ans;
    }

private:
    bool remainLengthValid(int octet, int remain) {
        return octet == 1 && remain >= 3 && remain <= 9 || 
            octet == 2 && remain >= 2 && remain <= 6 ||
            octet == 3 && remain >= 1 && remain <= 3 ||
            octet == 4 && remain == 0;
    }

    void dfs(int octet, string stack, string s, vector<string>& ans) {
        if (octet > 4) {
            if (s.empty()) {
                ans.push_back(stack.substr(0, stack.size() - 1));
            }
            return;
        }

        for (int l = 1; l <= 3; ++l) {
            if (!remainLengthValid(octet, s.size() - l)) {
                continue;
            }
            string prefix = s.substr(0, l);
            if (prefix[0] == '0' and l > 1) {
                return;
            }
            int value = stoi(prefix);
            if (value == 0) {
                dfs(octet + 1, stack + prefix + ".", s.substr(l), ans);
                return;
            }
            if (value > 0 && value < 256) {
                dfs(octet + 1, stack + prefix + ".", s.substr(l), ans);
            }
        }
    }
};

void print(vector<string> vec) {
    for (string s : vec) {
        cout << s << "  ";
    }
    cout << "\n";
}

int main() {
    Solution* solution = new Solution();
    vector<string> ans;

    print(solution->restoreIpAddresses("25525511135"));
    print(solution->restoreIpAddresses("0000"));
    print(solution->restoreIpAddresses("1111"));
    print(solution->restoreIpAddresses("010010"));
    print(solution->restoreIpAddresses("101023"));

}

